src/import-shadow-hash.c

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ 
+ 
+#include "storage.h"
+#include "module.h"
+#include "entry.h"
+#include <string.h>
+#include <stdio.h>
+#include <shadow.h>
+
+#define SH_TMPFILE "/etc/nshadow"
```

Comment:
If "/etc/shadow" is provided by a flag, then the tempfile name should be based off of that. I'm a fan of making the mktemp filename pattern something like "/etc/.shadow.tmp.XXXXXX". g_path_get_basename and g_path_get_dirname are handy for splitting up path parts (basename/dirname modify the orig string).

You may be able to not need this at all if you switch to "passwd --lock" and "--unlock" for making modifications.